### PR TITLE
v0.4.1 bugfix RC

### PR DESCRIPTION
**Pull request by josh146**
**Context:** v0.4.1 bugfix PR

**Description of the Change:**

Incorporates the following PRs: #439 #457 #437 #465 #471

### CODE REVIEW

File: doc/changelog.md [hunk 1]
```diff
@@ -1,4 +1,37 @@
...
+          qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+  
+      for_loop(0, 4, 1)(cnot_loop)()
+  
+      return qml.expval(qml.PauliZ(0))
+  ```
+
+* Resolve a memory leak in the runtime stemming from  missing calls to device destructors
+  at the end of programs.
+  [(#446)](https://github.com/PennyLaneAI/catalyst/pull/446)
```
Comment (josh146) [inline doc/changelog.md:27]: @maliasadi @dime10 I only cherry picked in commit 8f66c16. Do you know why this PR is also being pulled in? Is it just in the changelog, or below in the logic as well?

**Comment by josh146** [comment]
[sc-55556]

**Comment by rmoyard** [comment]
You can use the three actions from your branch

**Comment by dime10** [comment]
Note that 457 is not a bugfix, and also introduced some issue on mac x86. Was it required as a dependency of other changes? If so, there is also a minimal bugfix that resolves the x86 issue.

**Comment by josh146** [comment]
@rmoyard: Thanks, have activated the action!

@dime10: unfortunately cherry pick insisted on pulling in #457 while cherry picking in #465, since #465 was implemented on top of the #457 changes. We have two options:

- remove #457 while retaining #465 in this branch
- add in the MacOS x86 bugfix you mention @dime10 

@maliasadi @dime10 what do you recommend here?

**Comment by josh146** [comment]
Note: the MacOS x86 action seems to fail quite early: https://github.com/PennyLaneAI/catalyst/actions/runs/7698133685.

Is this because of the changes here? https://github.com/PennyLaneAI/catalyst/pull/409

**Comment by dime10** [comment]
Hi @josh146, the fix for the mac x86 action is here: [Disable LightningKokkos OMP backend on macOS (x86_64) by maliasadi · Pull Request #478 · PennyLaneAI/catalyst (github.com)](https://github.com/PennyLaneAI/catalyst/pull/478)

Additionally, I went through everything that was merged and might suggest including the following (if possible):
- [Address the issue of missing calls to device destructors at the end of programs by maliasadi · Pull Request #446 · PennyLaneAI/catalyst (github.com)](https://github.com/PennyLaneAI/catalyst/pull/446)
-  [Avoid infinite recursion in decomposition of `Controlled` by dime10 · Pull Request #468 · PennyLaneAI/catalyst (github.com)](https://github.com/PennyLaneAI/catalyst/pull/468)
- [Fix the issue with `AllocateRange` in the runtime by maliasadi · Pull Request #473 · PennyLaneAI/catalyst (github.com)](https://github.com/PennyLaneAI/catalyst/pull/473) (not a large priority)
- [Update sharp_bits.rst by josh146 · Pull Request #445 · PennyLaneAI/catalyst (github.com)](https://github.com/PennyLaneAI/catalyst/pull/445) (docs correction)

**Comment by dime10** [comment]
> Is this because of the changes here? https://github.com/PennyLaneAI/catalyst/pull/409

Yes it looks like you'll need Rashid's new runner as well

**Comment by josh146** [comment]
Thanks @dime10!

> Yes it looks like you'll need Rashid's new runner as well

Cool, I've just included #409 in this PR as well.

> https://github.com/PennyLaneAI/catalyst/pull/445 (docs correction)

This already appears to be in the most recent release: https://docs.pennylane.ai/projects/catalyst/en/stable/dev/sharp_bits.html#dynamically-shaped-arrays

> https://github.com/PennyLaneAI/catalyst/pull/446 (was accidentally merged into rc after release rather than main, hence you may have missed it)

Unfortunately I can't cherry pick this commit in, I think partially due to it not being merged into `main`, but instead entering `main` with the rest of the `rc` changes. Let's leave this one to the 0.5 release, unless it causes major problems.

File: .github/workflows/build-wheel-macos-arm64.yaml [hunk 2]
```diff
@@ -32,12 +32,6 @@ jobs:
     steps:
     - name: Checkout Catalyst repo
       uses: actions/checkout@v3
-    
-    - name: Setup Runner Environment
```
Comment (maliasadi) [inline .github/workflows/build-wheel-macos-arm64.yaml]: this should not be removed if you have included #409.

File: doc/changelog.md [hunk 3]
```diff
@@ -1,3 +1,55 @@
+# Release 0.4.1
+
+<h3>Improvements</h3>
+
+* Catalyst wheels are now packaged with OpenMP and ZStd.
+  [(#457)](https://github.com/PennyLaneAI/catalyst/pull/457)
+
+<h3>Bug fixes</h3>
```
Comment (maliasadi) [inline doc/changelog.md:8]: Looking into the main [changelog](https://github.com/PennyLaneAI/catalyst/blob/main/doc/changelog.md), references to the following PRs are missing,
- https://github.com/PennyLaneAI/catalyst/pull/409
- https://github.com/PennyLaneAI/catalyst/pull/446
- https://github.com/PennyLaneAI/catalyst/pull/473

File: doc/changelog.md [hunk 4]
```diff
@@ -1,3 +1,55 @@
...
+  def f(x):
+      def cnot_loop(j):
+          qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+  
+      for_loop(0, 4, 1)(cnot_loop)()
+  
+      return qml.expval(qml.PauliZ(0))
+  ```
+
+* Resolve unpredictable behaviour when importing libraries that share Catalyst's LLVM dependency
```
Comment (maliasadi) [inline doc/changelog.md:34]:
```suggestion

* Resolve a memory leak in the runtime stemming from  missing calls to device destructors
  at the end of programs.
  [(#446)](https://github.com/PennyLaneAI/catalyst/pull/446)

* Resolve unpredictable behaviour when importing libraries that share Catalyst's LLVM dependency
```

File: doc/changelog.md [hunk 5]
```diff
@@ -1,3 +1,55 @@
...
+  libraries can lead to process crashes and other unpredictable behaviour, since the wrong functions
+  can be called if both libraries are loaded in the current process.
+  The fix involves building shared libraries with hidden (macOS) or protected (linux) symbol
+  visibility by default, exporting only what is necessary.
+  [(#465)](https://github.com/PennyLaneAI/catalyst/pull/465)
+
+* Resolve a failure to find the SciPy OpenBLAS library when running Catalyst,
+  due to a different SciPy version being used to build Catalyst than to run it.
+  [(#471)](https://github.com/PennyLaneAI/catalyst/pull/471)
+
```
Comment (maliasadi) [inline doc/changelog.md:45]:
```suggestion

* Fix the issue in `LightningKokkos::AllocateQubits` with allocating too many qubit IDs on
  qubit re-allocation.
  [(#473)](https://github.com/PennyLaneAI/catalyst/pull/473)

```

File: doc/changelog.md (hunk 3 above)
Comment (maliasadi) [reply doc/changelog.md:8]: added them as code suggestions to this PR

**Comment by maliasadi** [comment]
> Hmm I think it was fixing some memory issues. How about cherry-picking the merge commit instead? https://github.com/PennyLaneAI/catalyst/pull/450

@josh146 I think there should not be any difference to cherry-picking #450 or #446 for this bugfix release. Compare to #446, it includes #444 (that's not related to the release) and the PR summary for #446 is also updated with a better wording in the changelog.

File: doc/changelog.md (hunk 4 above)
Comment (josh146) [reply doc/changelog.md:34]: @maliasadi I didn't include this PR, how come you are suggesting adding it to the changelog?

File: doc/changelog.md (hunk 5 above)
Comment (josh146) [reply doc/changelog.md:45]: @maliasadi I don't believe I included this PR either

File: .github/workflows/build-wheel-macos-arm64.yaml (hunk 2 above)
Comment (josh146) [reply .github/workflows/build-wheel-macos-arm64.yaml]: @maliasadi I think you are commenting on an old commit!

**Comment by josh146** [comment]
@dime10 @maliasadi I've included #446. Note that MacOS x86 wheels are building, but tests are failing: https://github.com/PennyLaneAI/catalyst/actions/runs/7698640517/job/20979540240

MacOS arm and linux are both passing.

**Comment by maliasadi** [comment]
@josh146 you also need to include https://github.com/PennyLaneAI/catalyst/pull/478 that includes the macOS x86 fix.

File: doc/changelog.md (hunk 4 above)
Comment (maliasadi) [reply doc/changelog.md:34]: nwm - I thought that you already included all David's suggestions https://github.com/PennyLaneAI/catalyst/pull/481#issuecomment-1914932651

File: doc/changelog.md (hunk 5 above)
Comment (maliasadi) [reply doc/changelog.md:45]: yup, you are right! will you also include that PR?

Comment (josh146) [reply doc/changelog.md:45]: Nope, I want to avoid pulling in too many PRs :)

File: doc/changelog.md [hunk 6]
```diff
@@ -1,3 +1,63 @@
+# Release 0.4.1
+
+<h3>Improvements</h3>
+
+* Catalyst wheels are now packaged with OpenMP and ZStd.
```
Comment (dime10) [inline doc/changelog.md:5]: This entry doesn't cover the "why" of the change / implications of the change.

**Review by dime10 — CHANGES_REQUESTED** [review]
Thanks @josh146, good work picking these commits from main!

File: doc/changelog.md [hunk 7]
```diff
@@ -1,3 +1,63 @@
...
+  due to a different SciPy version being used to build Catalyst than to run it.
+  [(#471)](https://github.com/PennyLaneAI/catalyst/pull/471)
+
+* Resolve a memory leak in the runtime stemming from  missing calls to device destructors
+  at the end of programs.
+  [(#446)](https://github.com/PennyLaneAI/catalyst/pull/446)
+
+* OpenMP support for the `lightning.kokkos` backend has been disabled on macOS x86_64,
+  due to memory issues in the computation of Lightning's adjoint-jacobian in the presence of multiple OMP threads.
+  [(#478)](https://github.com/PennyLaneAI/catalyst/pull/478)
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/changelog.md]: @dime10 @maliasadi done in 4699418 (huh, a commit has with no letters!)

**Comment by josh146** [comment]
- [macOS ARM wheels passing](https://github.com/PennyLaneAI/catalyst/actions/runs/7700980979) ✅ 
- [Linux wheels passing](https://github.com/PennyLaneAI/catalyst/actions/runs/7700980073) ✅ 
- [macOS x86 wheels passing](https://github.com/PennyLaneAI/catalyst/actions/runs/7700981904) ✅

**Review by dime10 — APPROVED** [review]
🚀

**Pull request merged**